larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h

```diff
@@ -0,0 +1,393 @@
+/**
+ *  @file   larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h
+ *
+ *  @brief  Header file for the lar discrete probability vector class
+ *
+ *  $Log: $
+ */
+#ifndef LAR_DISCRETE_PROBABILITY_VECTOR_H
+#define LAR_DISCRETE_PROBABILITY_VECTOR_H 1
+
+#include "Pandora/StatusCodes.h"
+
+#include <limits>
+#include <random>
+#include <utility>
+#include <vector>
+
+namespace lar_content
+{
+
+/**
+ *  @brief  DiscreteProbabilityVector class
+ */
+class DiscreteProbabilityVector
+{
+public:
+
+    template <typename TX, typename TY>
+    using InputDatum = std::pair<TX, TY>;
+
+    template <typename TX, typename TY>
+    using InputData = std::vector<InputDatum<TX, TY> >;
+
+    typedef std::vector<float> ResamplingPoints;
+
+    /**
+     *  @brief  Constructor
+     *
+     *  @param  inputData the data used to construt the probability vector
+     *  @param  xUpperBound the upper bound of the probability vector
+     *  @param  useWidths bool controlling whether the bin widths are used in calculations
+     */
+    template <typename TX, typename TY>
+    DiscreteProbabilityVector(const InputData<TX, TY> &inputData, const TX xUpperBound, const bool useWidths);
+
+    /**
+     *  @brief  Constructor
+     *
+     *  @param  discreteProbabilityVector a discrete probability vector to resample from
+     *  @param  resamplingPoints the points to resample the discrete probability vector with 
+     */
+    DiscreteProbabilityVector(const DiscreteProbabilityVector &discreteProbabilityVector, const ResamplingPoints &resamplingPoints);
+
+    /**
+     *  @brief  Constructor
+     *
+     *  @param  discreteProbabilityVector a discrete probability vector to randomly rearrange
+     *  @param  randomNumberGenerator the random number generator for the random reshuffling
+     */
+    DiscreteProbabilityVector(const DiscreteProbabilityVector &discreteProbabilityVector, std::mt19937 &randomNumberGenerator);
+
+    /**
+     *  @brief  Evaluate the cumulative probability at arbitrary x
+     *
+     *  @param  x the x value
+     *
+     *  @return the cumulative probability
+     */
+    float EvaluateCumulativeProbability(const float x) const;
+
+    /**
+     *  @brief  Get the size of the probability vector
+     *
+     *  @return the probability vector size
+     */
+    size_t GetSize() const;
```

Comment:
size_t vs. unsigned int discussion. Only advantage of using unsigned int seems to be consistency with existing. Maybe we can say that either is fine, so long as we're consistent within context of each class?

Reply:
Fixed in a previous commit.  Opted for unsigned int all-round